web/app/components/comment/comment.tsx

```diff
@@ -0,0 +1,739 @@
+  collapseToggle?: (id: CommentType['id']) => void;
+  setPinState?(id: CommentType['id'], value: boolean): Promise<void>;
+  blockUser?(id: User['id'], name: User['name'], ttl: BlockTTL): Promise<void>;
+  unblockUser?(id: User['id']): Promise<void>;
+  setVerifyStatus?(id: User['id'], value: boolean): Promise<void>;
+}
+
+export interface State {
+  isCopied: boolean;
+  isReplying: boolean;
+  isEditing: boolean;
+  editDeadline: Date | null;
+  voteErrorMessage: string | null;
+  /**
+   * delta of the score:
+   * default is 0.
+   * if user upvoted delta will be incremented
+   * if downvoted delta will be decremented
+   */
+  scoreDelta: number;
+  /**
+   * score copied from props, that updates instantly,
+   * without server response
+   */
+  cachedScore: number;
+}
+
+/**
+ * handler singletone which toggles replying/editing
+ * state between multiple comments
+ */
+let onPrevInputToggleCb: (() => void) | null = null;
+
+export class Comment extends Component<Props, State> {
+  votingPromise: Promise<unknown>;
+  /** comment text node. Used in comment text copying */
+  textNode?: HTMLDivElement;
+
+  constructor(props: Props) {
+    super(props);
+
+    this.state = {
+      isCopied: false,
+      isReplying: false,
+      isEditing: false,
+      editDeadline: null,
+      voteErrorMessage: null,
+      scoreDelta: 0,
+      cachedScore: props.data.score,
+    };
+
+    this.votingPromise = Promise.resolve();
+
+    this.updateState(props);
+
+    this.toggleEditing = this.toggleEditing.bind(this);
+    this.toggleReplying = this.toggleReplying.bind(this);
+    this.blockUser = debounce(this.blockUser, 100).bind(this);
+  }
+
+  componentWillReceiveProps(nextProps: Props) {
+    this.updateState(nextProps);
+  }
+
+  updateState(props: Props) {
+    let scoreDelta = 0;
+    if (props.user) {
+      if (props.data.votes[props.user.id] === true) {
+        ++scoreDelta;
+      }
+      if (props.data.votes[props.user.id] === false) {
+        --scoreDelta;
+      }
+    }
+
+    this.setState({
+      scoreDelta,
+      cachedScore: props.data.score,
+    });
+
+    if (props.user) {
+      const userId = props.user!.id;
+
+      // set comment edit timer
+      if (userId === props.data.user.id) {
+        const editDuration = StaticStore.config.edit_duration;
+        const timeDiff = StaticStore.serverClientTimeDiff || 0;
+        let editDeadline: Date | null = new Date(new Date(props.data.time).getTime() + timeDiff + editDuration * 1000);
+        if (editDeadline < new Date()) editDeadline = null;
+        this.setState({
+          editDeadline,
+        });
+      }
+    }
+  }
+
+  toggleReplying() {
+    const { isReplying } = this.state;
+    this.setState({ isEditing: false }, () => this.setState({ isReplying: !isReplying }));
+
+    if (onPrevInputToggleCb) onPrevInputToggleCb();
+
+    if (!isReplying) {
+      onPrevInputToggleCb = () => this.setState({ isReplying: false });
```

Comment:
Ouch, I believe it works, but I really don't like it

Reply:
Well, not my invention, but I find that it indeed works good, and don't add much cognitive overload. I'm as well don't like singletonish behaviour. Probably should change to something over redux.

Reply:
Edited in https://github.com/umputun/remark/pull/281/commits/a9f2f185a421e3d11f609e9eb27625279f8d803f. Looks a bit like overengineering :-) check it out.